README.md
predictiveFeatures Package Introduction
================
2022-12-15

**predictiveFeatures** is a package for the extraction of comprehensive features on genomic intervals. A fundamental task in a genomic data science project is to extract informative genomic metrics that can predict quantities defined on range-based genomic annotations. In the past, the feature engineering tasks are often handled by a small number of handcrafted genome-derived features and sequence features. However, these methods cannot fully explore the interactive space between genomic regions and different genomic properties, such as the length and sequence contents.

The methods implemented in the predictiveFeatures package can extract a wide range of properties defined on genomic regions, including length, sequence contents, genomic scores, clustering effects, distance toward the 5'/3' ends, and the relative positions of annotation on regions. When extracting genomic features with its main function, various genome properties are interactively extracted from the genomic regions of exons, introns, genes, transcripts, promoters, 5'UTR, 3'UTR, and CDS, establishing a large collection of genome-derived feature set. The input data for predictiveFeatures are any target genome annotation stored in a Granges object. For example, the target can be intervals of peaks or sites obtained by high-throughput sequencing assays (such as par-CLIP, iCLIP, ChIP-Seq, and meRIP-Seq).

Another important question in the analysis of functional genomic data is to identify the genomic factors that are highly correlated to the target, as these factors may be causally linked to the measurement of interest. predictiveFeatures can also offer a catalog of highly interpretive genomic & sequence features, which helps to identify the biologically meaningful factors through the feature importance analysis of the predictive models.

## Overview of functionalities

*predictiveFeatures* provides fast and comprehensive feature extraction from the
range-based genome annotation; the extracted features can be used as the
input for a variety of modeling purposes such as supervised predictive
modeling and unsupervised factor analysis. The target range-based
annotation should be stored in *GRanges* object, which can be created
using the constructor function defined in the *GenomicFeatures* package.
Alternatively, the *GRanges* can be imported from the external
annotation files in **BED**, **GFF**, or **GTF** formats using the
`import` function defined in *dplyr* package.

There are 2 types of genomic features implemented in *predictiveFeatures*: the
region properties features, and the sequence-derived features. The
former is extracted using an interactive design between the genomic
properties and the genomic regions, the latter is the sequence-derived
features extracted from the sequences of the flanking regions of the
annotation. When extracting the sequence-derived features, different
encoding methods can be specified, such as the one-hot method or the
pseudo nucleotide composition method.

<div class="figure" style="text-align: left">

<img src="inst/figures/predictiveFeatures_v1.0.png" alt="Feature extraction framework of predictiveFeatures" width="100%" />

<p class="caption">

Feature extraction framework of predictiveFeatures

</p>

</div>

## Download

Installation from github can be initiated with the following command:

``` r
devtools::install_github("zw-xjtlu/predictiveFeatures")
```

Alternatively, it can be downloaded from bioconductor with:

``` r
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("predictiveFeatures")
```

## Quick start

Here we demonstrate the most basic function of gnomic feature extraction. There are variety of bioconductor objects that are necessary to provide annotations if the corresponding property types are required, which we will display in the section bellow. The code chunk below assumes you have a GRanges for your target genomic intervals called `grg`. The `txdb` is a TxDb object for the transcript annotation, here we should use the annotation from the same genome assembly of `grg`. The `bsgemone` and `phastcons` are the reference genome and genome conservation scores, respectively. 


```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(predictiveFeatures)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(BSgenome.Hsapiens.UCSC.hg19)
library(phastCons100way.UCSC.hg19)

txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene 
bsgenome <- BSgenome.Hsapiens.UCSC.hg19
phastcons <- phastCons100way.UCSC.hg19 

set.seed(737)
grg <- GRanges(seqnames = rep(c("chr1", "chr2"), c(5, 15)),
               ranges = IRanges(c(sample(11874:12127, 5), 
                                  sample(38814:41527, 15)), 
                                width=1),
               strand = Rle(c("+", "-"), c(5, 15)))
              
gfeatures <- genomeDerivedFeatures(x = grg,
                                   transcriptdb = txdb,
                                   sequence = bsgenome,
                                   gscores = phastcons)
                                   
str(gfeatures) #Display the columns (features)
```

The following additional functions will be explained in the user's guide:

- If you want to extract features from DNA sequences (with *one-hot encoding*, *PseTNC*, ect.) you could store the regions of the DNA sequences into GRanges, and then you can use `sequenceDerivedFeatures()`.

- If you want to introduce mutations such as SNPs into sequence-derived features, which is often used by in-silico mutation analysis, you can use the `mutation` argument in `sequenceDerivedFeatures()`.

- If you want to define other properties over genomic regions, you can use `extractRegionProperty()`.




NEWS
CHANGES IN VERSION 1.0.0
------------------------

    o Workflow: The wrapper function genomeDerivedFeatures() 
      could annotate >200 genomic features given TxDb or EnsDb, 
      BSgenome, and GScores objects. Individual functions to create 
      Genomic features using other metric & region combinations 
      are still accessible.

    o Messages: The message generated by genomeDerivedFeatures()
      will display the feature name and information. Numeric indexes are also added at 
      the beginning of each feature. 


inst/tests/tests.R
library(GenomicRanges)
library(testthat)
library(predictiveFeatures)

x_gr <- GRanges(rep(c("chr1", "chr2"), c(5, 15)),
                IRanges(c(sample(11874:12127, 5), sample(38814:41527,15)), width=100),
                strand=Rle(c("+", "-"), c(5, 15)))

exons_gr <- GRanges(c("chr1","chr2","chr2"),
                    IRanges(start=c(11874,38814,45440),end=c(12227,41627,46588)),
                    strand=c("+","-","-"))
genes_grl <- GRangesList(gene1=exons_gr[1],gene2=exons_gr[c(2,3)])

y_gr <- GRanges(rep(c("chr1", "chr2"), c(50, 50)),
                IRanges(c(sample(11874:12127, 50),
                          sample(38814:41527,50)), width=1),
                strand=Rle(c("+", "-"), c(50, 50)))

expect_that(extractRegionLength(x_gr), is_a("integer"))
expect_that(extractRegionLength(x_gr, exons_gr), is_a("numeric"))
expect_that(extractRegionLength(x_gr, exons_gr), is_a("numeric"))
exons_property <- c(1,6,8)
expect_that(extractRegionProperty(x_gr, exons_gr, exons_property), is_a("numeric"))
expect_that(extractRegionYCount(x_gr, x_gr, exons_gr), is_a("numeric"))
expect_that(extractRegionNearestDistToY(x_gr, y_gr), is_a("numeric"))
expect_that(extractDistToRegion5end(x_gr, exons_gr), is_a("numeric"))
expect_that(extractDistToRegion3end(x_gr, exons_gr), is_a("numeric"))


R/AllGenerics.R
#' Method extractRegionProperty
#' 
#' @param x A \link{GRanges} object for the query.
#' @param region A \link{GRanges} or \link{GRangesList} object for the subject. 
#' If not provided (\code{region=NULL}), the returned properties will be calculated directly from \code{x}.
#' @param property A vector specifying the properties/attributes of the \code{region}. 
#' 
#' @param ambiguityMethod By default, for \code{logical} \code{property} input, if x overlaps multiple regions, 
#' as long as any of the properties is TRUE, the returned value will be TRUE. 
#' For other types of \code{property} input, the returned value will be the average of the properties over multiple overlapped regions.
#' If \code{ambiguityMethod} is \code{"mean"}, \code{"sum"}, \code{"min"}, or \code{"max"},
#' then the mean, sum, minimum, and maximum values of the >1 mapping will be returned in the output value.
#' 
#' @param maxgap,minoverlap,type See \code{?\link{findOverlaps}} in the \bold{IRanges} package for a description of these arguments. 
#' 
#' @param nomapValue When \code{nomapValue} is \code{"NA"}, \code{"0"}, or \code{"FALSE"}, 
#' the \code{x} that do not match the \code{region} will return \code{NA}, \code{0}, and \code{FALSE} respectively.
#' If \code{nomapValue} is \code{"nearest"}, the not matched \code{x} will be set to be the property value on its nearest \code{region}.
#'
#' @param sequence A \link{BSgenome} or \link{XStringSet} object. 
#' See the \link{available.genomes} function for how to install a genome.
#' 
#' @param gscores A \link{GScores} object. 
#' See the vignette for more details: \code{vignette("GenomicScores", package = "GenomicScores")}.
#' 
#' @param as.prob,letters See \code{?\link{letterFrequency}} in the \bold{Biostrings} package for a description of these arguments.
#'
#' @param ignore.strand When set to \code{TRUE}, the strand information is ignored in the overlap calculations.
#' 
#' @param output.logical If \code{TRUE} then the returned value is \code{logical}, otherwise \code{numeric}.
#' 
#' @param efficient \code{TRUE} if only internally extract the properties on \code{regions} overlap with \code{x}, 
#' the option makes the computation more efficient if the number of gnomic regions is much larger than the query.
#' 
#' @param missingScores Specifying the imputation methods on missing scores in \code{extractRegionScores}, 
#' default is returning zero for unknown gscores under the region.
#' 
#' @param y A \link{GRanges} or \link{GRangesList} object for the clustering annotation.
#' 
#' @param normalize If \code{TRUE}, then returned count in \code{extractRegionYCount} will be divided by the length of the region.
#' The normalized count can be understood as the density of the annotation on the area.
#' 
#' @param maxDist A value of the maximum nucleotide distance returned in \code{extractRegionNearestDistToY}, default: \code{3e+06}.
#' 
#' @param ... Additional arguments, it passes to \link{letterFrequency} in method \code{extractRegionLetterFrequency}, and it passes to \link{score} in method \code{extractRegionScores}.
#' 
#' @rdname extractRegionProperty-methods
#' @details Please check the documentation of \link{extractRegionProperty} for details.
#' @exportMethod extractRegionProperty
setGeneric("extractRegionProperty", function(x, 
                                             region, 
                                             property,
                                             ambiguityMethod = c("auto", "mean", "sum", "min", "max"),
                                             maxgap=0L, 
                                             minoverlap=0L,
                                             type=c("any", "start", "end", "within", "equal"),
                                             nomapValue=c("NA","0","FALSE","nearest"),
                                             ignore.strand=FALSE){standardGeneric("extractRegionProperty")})

#' Method extractRegionOverlap
#' @rdname extractRegionProperty-methods
#' @exportMethod extractRegionOverlap
setGeneric("extractRegionOverlap", function(x,
                                            region=NULL,
                                            ambiguityMethod=c("auto", "mean", "sum", "min", "max"),
                                            maxgap=-1L, 
                                            minoverlap=0L,
                                            type=c("any", "start", "end", "within", "equal"),
                                            ignore.strand=FALSE,
                                            output.logical=TRUE){standardGeneric("extractRegionOverlap")})

#' Method extractRegionLength
#' @rdname extractRegionProperty-methods
#' @exportMethod extractRegionLength
setGeneric("extractRegionLength", function(x,
                                           region=NULL,
                                           ambiguityMethod=c("mean", "sum", "min", "max"),
                                           maxgap=-1L, 
                                           minoverlap=0L,
                                           type=c("any", "start", "end", "within", "equal"),
                                           nomapValue=c("NA","0","nearest"),
                                           ignore.strand=FALSE){standardGeneric("extractRegionLength")})

#' Method extractRegionLetterFrequency
#' @rdname extractRegionProperty-methods
#' @exportMethod extractRegionLetterFrequency
setGeneric("extractRegionLetterFrequency", function(x,
                                                    sequence,
                                                    region=NULL,
                                                    letters = "GC",
                                                    as.prob = TRUE,
                                                    ambiguityMethod=c("mean", "sum", "min", "max"),
                                                    maxgap=-1L, 
                                                    minoverlap=0L,
                                                    type=c("any", "start", "end", "within", "equal"),
                                                    nomapValue=c("NA","0","nearest"),
                                                    ignore.strand=FALSE,
                                                    efficient=TRUE,
                                                    ...){standardGeneric("extractRegionLetterFrequency")})

#' Method extractRegionScores
#' @rdname extractRegionProperty-methods
#' @exportMethod extractRegionScores
setGeneric("extractRegionScores", function(x,
                                           gscores,
                                           region=NULL,
                                           ambiguityMethod=c("mean", "sum", "min", "max"),
                                           maxgap=-1L, 
                                           minoverlap=0L,
                                           type=c("any", "start", "end", "within", "equal"),
                                           nomapValue=c("NA","0","nearest"),
                                           missingScores=c("zero", "mean", "none"),
                                           ignore.strand=FALSE,
                                           efficient=TRUE,
                                           ...){standardGeneric("extractRegionScores")})


#' Method extractRegionYCount
#' @rdname extractRegionProperty-methods
#' @exportMethod extractRegionYCount
setGeneric("extractRegionYCount", function(x,
                                           y,
                                           region=NULL,
                                           normalize=FALSE,
                                           ambiguityMethod=c("mean", "sum", "min", "max"),
                                           maxgap=-1L, 
                                           minoverlap=0L,
                                           type=c("any", "start", "end", "within", "equal"),
                                           nomapValue=c("NA","0","nearest"),
                                           ignore.strand=FALSE,
                                           efficient=TRUE){standardGeneric("extractRegionYCount")})


#' Method extractRegionNearestDistToY
#' @rdname extractRegionProperty-methods
#' @exportMethod extractRegionNearestDistToY
setGeneric("extractRegionNearestDistToY", function(x,
                                                   y,
                                                   region=NULL,
                                                   ambiguityMethod=c("mean", "sum", "min", "max"),
                                                   maxgap=-1L, 
                                                   minoverlap=0L,
                                                   type=c("any", "start", "end", "within", "equal"),
                                                   nomapValue=c("NA","0","nearest"),
                                                   maxDist=3e6,
                                                   ignore.strand=FALSE){standardGeneric("extractRegionNearestDistToY")})

#' Method extractRegionRelativePosition
#' @rdname extractRegionProperty-methods
#' @exportMethod extractRegionRelativePosition
setGeneric("extractRegionRelativePosition", function(x,
                                                     region=NULL,
                                                     ambiguityMethod=c("mean", "sum", "min", "max"),
                                                     nomapValue=c("NA","0"),
                                                     ignore.strand=FALSE){standardGeneric("extractRegionRelativePosition")})


#' Method extractDistToRegion5end
#' @rdname extractRegionProperty-methods
#' @exportMethod extractDistToRegion5end
setGeneric("extractDistToRegion5end", function(x,
                                               region=NULL,
                                               ignore.strand=FALSE,
                                               ambiguityMethod=c("mean", "sum", "min", "max"),
                                               maxgap=-1L, 
                                               minoverlap=0L,
                                               type=c("any", "start", "end", "within", "equal"),
                                               nomapValue=c("NA","0","nearest")){standardGeneric("extractDistToRegion5end")})

#' Method extractDistToRegion3end
#' @rdname extractRegionProperty-methods
#' @exportMethod extractDistToRegion3end
setGeneric("extractDistToRegion3end", function(x,
                                               region=NULL,
                                               ignore.strand=FALSE,
                                               ambiguityMethod=c("mean", "sum", "min", "max"),
                                               maxgap=-1L, 
                                               minoverlap=0L,
                                               type=c("any", "start", "end", "within", "equal"),
                                               nomapValue=c("NA","0","nearest")){standardGeneric("extractDistToRegion3end")})


R/sampleSequence.R
#' @title Sample nucleotides or motifs on a subset of genome defined by \code{GRanges}.
#'
#' @description A function used to extract the regions of a query sequence mapped to user defined sub-regions of the genome.
#'
#' @param motif a character string indicating the query sequence or motifs; Vague mapping rules in \code{\link{IUPAC_CODE_MAP}} is supported when \code{Fixed} = FALSE.
#'
#' @param region the \code{\link{GRanges}} object used to define the subset region of the genome.
#'
#' @param sequence the \code{\link{BSgenome}} object containing the sequence of the genome.
#'
#' @param fixed FALSE to support the vague mapping rule of the character string, default is FALSE.
#'
#' @param N number of ranges sampled, by default, it returns all the matched ranges without sub-sampling.
#'
#' @param replace whether sample with replacement or not, default is FALSE.
#'
#' @return a \code{GRanges} object contains the (sampled) mapped regions of the query sequence on the given subset of the genome.
#'
#' @examples
#' ## Retrieve the dimer sequence "CG" on 3 exons of hg19
#' library(TxDb.Hsapiens.UCSC.hg19.knownGene)
#' library(BSgenome.Hsapiens.UCSC.hg19)
#' exons_gr <- GRanges(c("chr1","chr2","chr2"),
#'                     IRanges(start=c(11874,38814,45440),
#'                     end=c(12227,41627,46588)),
#'                     strand=c("+","-","-"))
#' CpG_regions <- sampleSequence("GG", exons_gr, Hsapiens)
#'
#' \donttest{
#' ## Sample 100000 ranges of DRACH motif (the motif of m6A) on all exon regions of hg19:
#' exons_hg19 <- exons(TxDb.Hsapiens.UCSC.hg19.knownGene)
#' Motif_exons_hg19 <- sampleSequence("DRACH", exons_hg19, Hsapiens, N = 100000)
#' }
#'
#' @importFrom GenomicRanges reduce GRanges
#' @importFrom GenomicFeatures mapFromTranscripts
#' @importFrom GenomeInfoDb seqlengths
#' @importFrom Biostrings vmatchPattern
#' @export
sampleSequence <- function(motif, region, sequence, N = NULL, fixed = FALSE, replace = FALSE){
  stopifnot(is(region, "GRangesList")|is(region, "GRanges"))
  if(is(region, "GRangesList")) region <- unlist(region)
  region <- reduce(region)

  region_dnass <- getSeq(x=sequence,
                         names=seqnames(region),
                         start=start(region),
                         end=end(region),
                         strand=strand(region),
                         as.character=FALSE)
  indx <- paste0("reg_", seq_along(region))
  regions_GRL <- split(region, indx)
  regions_GRL <- regions_GRL[indx]
  rm(indx)
  vmp <- vmatchPattern(motif, region_dnass, fixed = fixed)
  rm(region_dnass)
  vmp_gr <- GRanges(seqnames = rep(names(regions_GRL), elementNROWS(vmp)), ranges = unlist(vmp))
  rm(vmp)
  motif_on_regions <- mapFromTranscripts(vmp_gr,regions_GRL)
  rm(vmp_gr, regions_GRL)
  mcols(motif_on_regions) = NULL
  if(is.null(N)) N <- length(motif_on_regions)
  if(replace == FALSE) N <- min(N, length(motif_on_regions))
  indx2 <- sample.int(length(motif_on_regions), N, replace = replace)
  motif_on_regions <- motif_on_regions[indx2]
  rm(indx2)
  seqlengths(motif_on_regions) <- seqlengths(region)
  return(motif_on_regions)
}


R/PackageDocumentation.R
#' predictiveFeatures: extract genomic features for predictive modeling in genomics
#'
#' The package can be used to extract properties of region lengths, sequence contents, relative positions, clustering effects, and conservation scores on top of the classically defined gene annotations including 5'UTR, CDS, 3'UTR, transcripts, genes, and promoters. The genome-derived features can be used to predict variety of gene related genomic assays, such as the RBP binding sites and the RNA modification sites. predictiveFeatures also provide the annotation of new attributes over user-defined genomic regions, enabling the augmentation of feature engineering under the interactive framework between genomic properties and genomic regions.
#' 
#' The package currently implements 2 types of feature extraction modules: the \strong{genome-derived features} and the \strong{sequence-derived features}.The former is encoded via the interaction between genomic properties and the genomic regions, and the later is defined through the nucleotide encoding methods, such as the one-hot encoding or the encoding of pseudo nucleotide compositions (PseTNC). The genome and sequence features can be extracted with the function \link{genomeDerivedFeatures} and \link{sequenceDerivedFeatures} respectively. Furthermore, more genome-derived features can be defined using functions with names "extractRegion" + the metric names. While the individual extractor functions can annotate new properties on the user defined region types.
#' 
#' The design of package enables the one-step extraction of hundreds of genomic features based only on the \code{GRanges} of the target genome intervals. Wide class of the Bioconductor annotation objects, including \code{Txdb}, \code{EnsDb}, \code{BSgenome}, and \code{GScores}, are supported in the package to enable the massive enumeration of features that based on novel sources of genome regions.
#' 
#' While the comprehensive feature input is vital to the success of a genomic data science project, the genome-derived features are often more informative and interpretive than the primary sequence features. Thus, the \code{predictiveFearures} package can provide general support for machine learning modeling in the domain of genomics.
#' 
#' The user's guide will firstly give an overview to the key functionalities of the package, then it will illustrate the utility of the predictive features using a case study of the full workflow in a genomic prediction task, from EDA to modeling.
#'
#' @docType package
#' @name predictiveFeatures
#' @section Key predictiveFeatures functions:
#' \link{genomeDerivedFeatures}
#' 
#' \link{sequenceDerivedFeatures}
#' 
#' \link{topologyOnTranscripts}
NULL
#> NULL

R/topologyOnTranscripts.R
#' @title Extracting the topology on transcripts
#' 
#' @description A function to extract meta-transcript topologies of the input GRanges object
#' 
#' @param x A \link{GRanges} object for the genomic ranges to be annotated.
#' @param txdb A \link{TxDb} or \link{EnsDb} object for the transcript annotation.
#' @param region_weights A numeric vector of length 3 indicating the weights for 5'UTR, CDS, and 3'UTR; default is \code{c(1/3,1/3,1/3)}.
#' @param ambiguityMethod If \code{ambiguityMethod} is \code{"mean"} (default), \code{"sum"}, \code{"min"}, or \code{"max"},
#' then the mean, sum, minimum, and maximum values of the multiple region mappings will be returned in the output value.
#' @param ignore.strand When set to \code{TRUE}, the strand information is ignored in the overlap calculations.
#' 
#' @return A numeric vector with the same length as x. 
#' 
#' @details The meta-tx topology is calculated based on the weighted sum of the relative position of \code{x} on exonic 5'UTR, CDS, and 3'UTR,
#' that is, the topology \code{= sum(c(rel_pos_5UTR, rel_pos_CDS, rel_pos_3UTR)*region_weights)}. 
#' The \code{rel_pos} for x instances not mapped to the region is equal to 1 or 0. Specifically,
#' \code{rel_pos} is 1 if \code{x} overlaps at a downstream region,
#' \code{rel_pos} is 0 if \code{x} overlaps at an upstream region. The topology values for \code{x} that is not mapped to any one of the 5'UTR, CDS, and 3'UTR will be set to \code{NA}.
#' 
#' The relative positions on a give region are directly extracted by the function \code{extractRegionRelativePosition()}.
#' 
#' @examples 
#' ## Load the TxDb object
#' library(TxDb.Hsapiens.UCSC.hg19.knownGene)
#' txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
#' 
#' ## Define the query GRanges
#' set.seed(01)
#' 
#' query_gr <- GRanges(rep(c("chr1", "chr2"), c(5, 15)),
#'                 IRanges(c(sample(11874:12127, 5), sample(38814:41527, 15)), width=1),
#'                 strand=Rle(c("+", "-"), c(5, 15)))
#'                 
#' ## Extract the meta-tx topology values
#' topologyOnTranscripts(query_gr, txdb)
#' 
#' ## Visualize the logistic regression curve with B-splines on a m6A miCLIP classification data set
#' 
#' GSE63753_sysy <- readRDS(system.file("extdata", "GSE63753_sysy.rds", package = "predictiveFeatures"))
#' 
#' GSE63753_sysy$topology <- topologyOnTranscripts(GSE63753_sysy, txdb)
#'
#' library(ggplot2)
#' 
#' ggplot(na.omit(as.data.frame(mcols(GSE63753_sysy))), aes(topology, target)) +
#'   geom_smooth(formula = y ~ splines::ns(x, 8), method = "glm", method.args = list(family = "binomial")) +
#'   geom_vline(xintercept = c(0.33, 0.66), linetype = 2) +
#'   geom_hline(yintercept = 0.5, linetype = 3) +
#'   geom_text(aes(x=x,y=y,label=text),
#'             data = data.frame(
#'                x = c(0.165, 0.495, 0.825),
#'                y = c(0.1, 0.1, 0.1),
#'                text = c("5'UTR","CDS", "3'UTR")
#'   )) +
#'   scale_x_continuous(breaks = c(0, 0.33, 0.66, 0.9)) + scale_y_continuous(breaks = c(0, 0.25, 0.5, 0.75, 1)) +
#'   theme_classic() + labs(x = "meta-tx topology", y = "prob of m6A = 1", title = "LR fit with cubic splines")
#' @export
topologyOnTranscripts <- function(x, 
                                  txdb, 
                                  region_weights = c(1/3,1/3,1/3),
                                  ambiguityMethod = c("mean", "sum", "min", "max"),
                                  ignore.strand=FALSE){
  
  u5bytx <- EnsureUCSC(fiveUTRsByTranscript(txdb),x)
  topology <- extractRegionRelativePosition(x,
                                          u5bytx,
                                          ambiguityMethod=ambiguityMethod,
                                          nomapValue="NA",
                                          ignore.strand=ignore.strand)*region_weights[1]
  rm(u5bytx)
  cdsbytx <- EnsureUCSC(cdsBy(txdb, by = "tx"),x)
  cdsrps <- extractRegionRelativePosition(x,
                                        cdsbytx,
                                        ambiguityMethod=ambiguityMethod,
                                        nomapValue="NA",
                                        ignore.strand=ignore.strand)
  rm(cdsbytx)
  indx <- !is.na(cdsrps)
  topology[indx] <- cdsrps[indx]*region_weights[2] + region_weights[1]
  rm(indx,cdsrps)
  
  u3bytx <- EnsureUCSC(threeUTRsByTranscript(txdb),x)
  u3rps <- extractRegionRelativePosition(x,
                                       u3bytx,
                                       ambiguityMethod=ambiguityMethod,
                                       nomapValue="NA",
                                       ignore.strand=ignore.strand)
  rm(u3bytx)
  indx <- !is.na(u3rps)
  topology[indx] <- u3rps[indx]*region_weights[3] + region_weights[2] + region_weights[1]
  rm(indx,u3rps)
  
  return(topology)
}
